odl/discr/discr_utils.py

```diff
@@ -817,6 +834,573 @@ def __init__(self, coord_vecs, values, input_type):
+
+    Also verify that the signature of ``func`` has no ``*args`` since
+    they make argument propagation a huge hassle.
+
+    Note: this function only works for objects that can be inspected
+    with the ``inspect`` module, i.e., Python functions and callables,
+    but not, e.g., NumPy UFuncs.
+
+    Parameters
+    ----------
+    func : callable
+        Object that should be inspected.
+
+    Returns
+    -------
+    has_out : bool
+        ``True`` if the signature has an ``out`` argument, ``False``
+        otherwise.
+    out_is_optional : bool
+        ``True`` if ``out`` is present and optional in the signature,
+        ``False`` otherwise.
+
+    Raises
+    ------
+    TypeError
+        If ``func``'s signature has ``*args``.
+    """
+    if sys.version_info.major > 2:
+        spec = inspect.getfullargspec(func)
+        kw_only = spec.kwonlyargs
+    else:
+        spec = inspect.getargspec(func)
+        kw_only = ()
+
+    if spec.varargs is not None:
+        raise TypeError('*args not allowed in function signature')
+
+    pos_args = spec.args
+    pos_defaults = () if spec.defaults is None else spec.defaults
+
+    if 'out' in pos_args:
+        has_out = True
+        out_optional = (
+            pos_args.index('out') >= len(pos_args) - len(pos_defaults)
+        )
+    elif 'out' in kw_only:
+        has_out = out_optional = True
+    else:
+        has_out = out_optional = False
+
+    return has_out, out_optional
+
+
+def _func_out_type(func):
+    """Determine the output argument type (if any) of a function-like object.
+
+    This function is intended to work with all types of callables
+    that are used as input to `sampling_function`.
+    """
+    # Numpy `UFuncs` and similar objects (e.g. Numba `DUFuncs`)
+    if hasattr(func, 'nin') and hasattr(func, 'nout'):
+        if func.nin != 1:
+            raise ValueError(
+                'ufunc {} takes {} input arguments, expected 1'
+                ''.format(func.__name__, func.nin)
+            )
+        if func.nout > 1:
+            raise ValueError(
+                'ufunc {} returns {} outputs, expected 0 or 1'
+                ''.format(func.__name__, func.nout)
+            )
+        has_out = out_optional = (func.nout == 1)
+    elif inspect.isfunction(func):
+        has_out, out_optional = _check_func_out_arg(func)
+    elif callable(func):
+        has_out, out_optional = _check_func_out_arg(func.__call__)
+    else:
+        raise TypeError('object {!r} not callable'.format(func))
+
+    return has_out, out_optional
+
+
+def sampling_function(func_or_arr, domain, out_dtype=None):
+    """Return a function that can be used for sampling.
+
+    For examples on this function's usage, see `point_collocation`.
+
+    Parameters
+    ----------
+    func_or_arr : callable or array-like
+        Either a single callable object (possibly with multiple output
+        components), or an array or callables and constants.
```

Comment:
```suggestion
        components), or an array or callables and constants.
        A callable (or each callable) must take a single input and may
        accept one output parameter called ``out``, and should return
        its result.
```
To address [this comment](https://github.com/odlgroup/odl/pull/1459/files/eaebdc6bf721345bc1614a756ed3e12264b3fbe0#r400704096)